content/blog/15-making-rust-attractive-for-writing-gtk-applications.md

```diff
@@ -0,0 +1,301 @@
+- Writing the necessary infrastructre code to create custom GObjects and so allowing the end user to make their own custom widgets or GStreamer plugins for example
+
+TODO: add some numbers about the amount of manual code in gtk4-rs
+
+### Subclassing
+
+In the early days of `gtk3-rs`, the infrastructure for writing custom subclassses wasn't fully there yet, especially that the amount of manual code that has to be written for supporting all the virtual functions (that can be overridden by a sub-implementation) of `gtk::Widget` was a lot. That caused people to avoid writing custom GTK widgets and do plenty of hacks like having one single `ObjectWrapper` GObject that would serialize a Rust struct into a `JSON` to store it as a string property in order to store these Rust types in a `gio::ListModel` and use it with `gtk::ListBox::bind_model`/`gtk::FlowBox::bind_model`.
+
+Thankfully that is no longer the case for `gtk4-rs` as a huge amount of work went into making sure we support almost all the types that can be subclassed in Rust with the exception of `gtk::TreeModel`, see <https://github.com/gtk-rs/gtk4-rs/pull/169> for details why that didn't happen yet.
+
+As more people started writing custom GTK widgets/GStreamer plugins, the amount of people looking into simplifying the whole experience grow up.
+
+Creating a very simple and useless custom GTK widget in the days of gtk-rs-core 0.10
+```rust
+mod imp {
+    pub struct SimpleWidget;
+
+    impl ObjectSubclass for SimpleWidget {
+        const NAME: &'static str = "SimpleWidget";
+        type ParentType = gtk::Widget;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+
+        glib_object_subclass!();
+
+        fn new() -> Self {
+            Self()
+        }
+    }
+
+    impl ObjectImpl for SimpleWidget {
+        glib_object_impl!();
+    }
+    impl WidgetImpl for SimpleWidget {}
+}
+glib::wrapper! {
+    pub struct SimpleWidget(ObjectSubclass<imp::SimpleWidget>)
+        @extends gtk::Widget;
+}
+```
+
+The same code nowadays
+```rust
+mod imp {
+    #[derive(Default)]
+    pub struct SimpleWidget;
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for SimpleWidget {
+        const NAME: &'static str = "SimpleWidget";
+        type ParentType = gtk::Widget;
+    }
+
+    impl ObjectImpl for SimpleWidget {}
+    impl WidgetImpl for SimpleWidget {}
+}
+glib::wrapper! {
+    pub struct SimpleWidget(ObjectSubclass<imp::SimpleWidget>)
+        @extends gtk::Widget;
+}
+```
+
+Of course, the overall code is still a bit too verbose particularly when you have to define `GObject` properties. Although, people have been experimenting with writing a [derive macro](https://github.com/gtk-rs/gtk-rs-core/pull/494) to simplify the properties declaration part as well as a [`gobject::class!`](https://github.com/gtk-rs/gtk-rs-core/issues/540) macro for generating most of the remaining boilerplate code.
```

Comment:
```suggestion
Of course, the overall code is still a bit too verbose particularly when you have to define `GObject` properties. However, people have been experimenting with writing a [derive macro](https://github.com/gtk-rs/gtk-rs-core/pull/494) to simplify the properties declaration part as well as a [`gobject::class!`](https://github.com/gtk-rs/gtk-rs-core/issues/540) macro for generating most of the remaining boilerplate code.
```

Maybe add a sentence that these are still experiments and needs some more experimenting and time to mature